File: 3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        map<char,int>mpp;

        int left=0,right=0,len=0;

        while(right<s.length())
        {
            mpp[s[right]]++;
            if(mpp.size()==right-left+1)
            {
                len=max(right-left+1,len);
            }
            else if(mpp.size()<right-left+1)
            {
                while(mpp.size()<right-left+1)
                {
                    mpp[s[left]]--;
                    if(mpp[s[left]]==0)
                    {
                        mpp.erase(s[left]);
                    }
                    left++;
                }
            }

            right++;
        }

        return len;
    }
};
// @lc code=end

File: 3.longest-substring-without-repeating-characters.java


import java.util.Arrays;

/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {

        int[]prev=new int[256];
        int res=0,l=0,r=0;
        Arrays.fill(prev, -1);

        while(r<s.length()){
            if(prev[s.charAt(r)]>=l){
                l=prev[s.charAt(r)]+1;
            }

            res=Math.max(res, r-l+1);
            prev[s.charAt(r)]=r;
            r++;
        }

        return res;
        
    }
}
// @lc code=end

